Gpio support v2

MathisDeroo opened the pull request:
Modified GPIO driver to support access to PortX and pcrX registers using paste!. 
Added support for pin configuration (slew rate, drive strength, mux...) available in the PAC. 
Added button example to demonstrate input feature.

MathisDeroo committed:
Improve GPIO driver and add button example
- port and pcr registers are defined using paste!
- added pin configuration for slew rate, drive strength, mux function, etc...
- added button example to showcase input gpio feature

Signed-off-by: Mathis Deroo <mathis.deroo@nxp.com>

MathisDeroo committed:
Add pull-up pull-down config support for input gpio

Signed-off-by: Mathis Deroo <mathis.deroo@nxp.com>

MathisDeroo committed:
Replace GPIOs enum with existing ones in the PAC

Signed-off-by: Mathis Deroo <mathis.deroo@nxp.com>

MathisDeroo committed:
Remove init_gpio_pin function as it is done in hal init config

Signed-off-by: Mathis Deroo <mathis.deroo@nxp.com>

felipebalbi commented on src/gpio.rs:
```diff
@@ -1,4 +1,4 @@
-//! GPIO driver built around a type-erased `Flex` pin, similar to other Embassy HALs.
+//! GP driver built around a type-erased `Flex` pin, similar to other Embassy HALs.
```
Looks like an unnecessary change. Could you remove it?

MathisDeroo replied:
Will remove it

felipebalbi commented on src/gpio.rs:
```diff
@@ -8,16 +8,28 @@ use core::marker::PhantomData;
 pub type Gpio = crate::peripherals::GPIO0;
 
-/// Type-erased representation of a GPIO pin.
+/// Type-erased representation of a GP pin.
```
also looks unnecessary.

MathisDeroo replied:
Will remove it

jamesmunns replied:
Still needs to be fixed

felipebalbi commented on src/gpio.rs:
```diff
@@ -272,30 +390,6 @@ impl_pin!(P4_4, 4, 4, Gpio4);
 impl_pin!(P4_5, 4, 5, Gpio4);
 impl_pin!(P4_6, 4, 6, Gpio4);
 impl_pin!(P4_7, 4, 7, Gpio4);
-impl_pin!(P4_8, 4, 8, Gpio4);
-impl_pin!(P4_9, 4, 9, Gpio4);
-impl_pin!(P4_10, 4, 10, Gpio4);
-impl_pin!(P4_11, 4, 11, Gpio4);
-impl_pin!(P4_12, 4, 12, Gpio4);
-impl_pin!(P4_13, 4, 13, Gpio4);
-impl_pin!(P4_14, 4, 14, Gpio4);
-impl_pin!(P4_15, 4, 15, Gpio4);
-impl_pin!(P4_16, 4, 16, Gpio4);
-impl_pin!(P4_17, 4, 17, Gpio4);
-impl_pin!(P4_18, 4, 18, Gpio4);
-impl_pin!(P4_19, 4, 19, Gpio4);
-impl_pin!(P4_20, 4, 20, Gpio4);
-impl_pin!(P4_21, 4, 21, Gpio4);
-impl_pin!(P4_22, 4, 22, Gpio4);
-impl_pin!(P4_23, 4, 23, Gpio4);
-impl_pin!(P4_24, 4, 24, Gpio4);
-impl_pin!(P4_25, 4, 25, Gpio4);
-impl_pin!(P4_26, 4, 26, Gpio4);
-impl_pin!(P4_27, 4, 27, Gpio4);
-impl_pin!(P4_28, 4, 28, Gpio4);
-impl_pin!(P4_29, 4, 29, Gpio4);
-impl_pin!(P4_30, 4, 30, Gpio4);
-impl_pin!(P4_31, 4, 31, Gpio4);
 
```
why delete so many pins?

MathisDeroo replied:
Per my understanding some pins are not available. Please refer to the schematic

jamesmunns replied:
Is this unavailable *on the chip*, or *on the board*?

In most HALs, we focus on whatever the chip can support. If not all chips have all pins (smaller pin count packages), we can have chip-specific features to handle this.

felipebalbi replied:
Right, remember the HAL is for the chip, not a specific board :-) We just happen to use one particular board for testing.

MathisDeroo replied:
Got it, will revert my changes 😄

felipebalbi commented on src/gpio.rs:
```diff
@@ -395,19 +499,43 @@ impl<'d> Flex<'d> {
+
+    pub fn get_level(&self) -> Level {
+        self.is_high().into()
+    }
```
I should have been a little more specific. When it comes to the function, that's usually handled within the HAL. User doesn't mux pins explicitly. Because of that, we don't need human-readable names for the enum. Drive strength, pull, slew rate, and level are all exposed to the user and human-readable enums and variants become useful.

felipebalbi commented on src/gpio.rs:
```diff
@@ -395,19 +499,43 @@ impl<'d> Flex<'d> {
 }
 
-/// GPIO output driver that owns a `Flex` pin.
+/// GP output driver that owns a `Flex` pin.
```
unnecessary change?

MathisDeroo replied:
Will remove it

felipebalbi commented on src/gpio.rs:
```diff
@@ -395,19 +499,43 @@ impl<'d> Flex<'d> {
         let mut flex = Flex::new(pin);
         flex.set_level(initial);
-        flex.set_as_output();
+        flex.set_as_output(strength, slew_rate);
```
I think I would prefer separate `set_strength` and `set_slew_rate` functions.

MathisDeroo replied:
Will add both functions

felipebalbi commented on src/gpio.rs:
```diff
@@ -454,16 +582,21 @@ impl<'d> Output<'d> {
 }
 
-/// GPIO input driver that owns a `Flex` pin.
+/// GP input driver that owns a `Flex` pin.
```
unnecessary change?

MathisDeroo replied:
Will remove it

felipebalbi commented on src/gpio.rs:
```diff
@@ -454,16 +582,21 @@ impl<'d> Output<'d> {
+        ) -> Self {
         let mut flex = Flex::new(pin);
-        flex.set_as_input();
+        flex.set_as_input(pull_enable, pull_select, strength, slew_rate);
```
also a `set_pull` would be useful.

MathisDeroo replied:
okay will add it

felipebalbi commented on src/lib.rs:
```diff
@@ -225,63 +218,6 @@ embassy_hal_internal::peripherals!(
-    P4_25,
-    P4_26,
-    P4_27,
-    P4_28,
-    P4_29,
-    P4_30,
-    P4_31,
-
-    P5_0,
-    P5_1,
-    P5_2,
-    P5_3,
-    P5_4,
-    P5_5,
-    P5_6,
-    P5_7,
-    P5_8,
-    P5_9,
-    P5_10,
-    P5_11,
-    P5_12,
-    P5_13,
-    P5_14,
-    P5_15,
-    P5_16,
-    P5_17,
-    P5_18,
-    P5_19,
-    P5_20,
-    P5_21,
-    P5_22,
-    P5_23,
-    P5_24,
-    P5_25,
-    P5_26,
-    P5_27,
-    P5_28,
-    P5_29,
-    P5_30,
-    P5_31,
```
I had added all peripherals and you accidentally removed several. Can you revert this hunk and recover the peripherals?

MathisDeroo replied:
Same comment as before. Some pins are not available so I removed them.

MathisDeroo committed:
Integrate feedback for the GPIO driver

- Add again missing IO peripherals
- Added function to configure separately slew rate, drive strength, pull.
- Revert comment changes

Signed-off-by: Mathis Deroo <mathis.deroo@nxp.com>

jamesmunns commented on src/gpio.rs:
```diff
@@ -80,52 +107,165 @@ pub trait GpioPin: SealedPin + Sized + PeripheralType + Into<AnyPin> + 'static {
+        }
+    }
+
+    fn set_drive_strength(&self, strength: Dse) {
```
As a note, we *usually* don't expose the PAC types in the HAL, to allow us to make breaking changes to the PAC without making breaking changes in the HAL.

It might be worth defining new enums for strength, slew, etc., so we can hide that impl detail.

The PAC values are okay to use for internal/private functions, just not public APIs!

MathisDeroo replied:
So in the Output, Input and Flex implementation i.e where the public APIs are, I should refer to an enum I need to create like:
#[derive(Copy, Clone, Eq, PartialEq, Debug)]
pub enum SlewRate {
    Fast,
    Slow,
}

impl From<SlewRate> for Sre {
    fn from(slew_rate: SlewRate) -> Self {
        match slew_rate {
            SlewRate::Fast => Sre::Sre0,
            SlewRate::Slow => Sre::Sre1,
        }
    }
}

So the user can use readable fields configuration without breaking the underlying HAL implementation ?

jamesmunns replied:
That is what is typically done, yes!

MathisDeroo replied:
Should I then use into() in the public API to match with the private one ? 
pub fn set_slew_rate(&mut self, slew_rate: SlewRate) {
        self.pin.set_slew_rate(slew_rate.into());
    }

Or in the private API, pass as parameter the same type:
fn set_slew_rate(&self, slew_rate: SlewRate) {
       self.pcr_reg().modify(|_, w| w.sre().variant(slew_rate));
    }

jamesmunns replied:
you can, or you can do it more manual without an into/from impl, like:

```rust
pub fn set_slew_rate(&mut self, slew_rate: SlewRate) {
  match slew_rate {
    SlewRate::Fast => self.pcr_reg().modify(|_r, w| w.sre().sre0()),
    SlewRate::Slow => self.pcr_reg().modify(|_r, w| w.sre().sre1()),
  }
  //
  // or
  //
  let rate = match slew_rate {
    SlewRate::Fast => Sre::Sre0,
    SlewRate::Slow => Sre::Sre1,
  };
  self.pcr_reg().modify(|_r, w| w.sre().variant(rate));
}
```

But you can also do it the way you wrote, whatever is most convenient for you :)

jamesmunns commented on src/gpio.rs:
```diff
@@ -80,52 +107,165 @@ pub trait GpioPin: SealedPin + Sized + PeripheralType + Into<AnyPin> + 'static {
+                fn set_pull(&self, pull_enable: Pe, pull_select: Ps) {
+                    let port_reg = unsafe {&*crate::pac::[<Port $port>]::ptr()};
+
+                    if pull_enable == Pe::Pe0 {
```
is `.pe()` missing a `.variant()` you could use here? Is this something we need to fix with the PAC?

MathisDeroo replied:
I could use a variant. If when the pull is disabled, assigning the pull "state" has no effect then I could only do 
port_reg.[<pcr $pin>]().modify(|_, w| {
                            w.pe().variant(pull_enable)
                            w.ps().variant(pull_select)
});

jamesmunns replied:
Ah, thanks! It might make sense to make the public type you use here something like:

```rust
enum Pull {
   Up,
   Down,
   Disabled,
}
```

And then do `set_pull(&self, pull: Pull)`.

I didn't totally understand what `Pe0` was when I was looking over this :D

jamesmunns commented on src/gpio.rs:
```diff
@@ -80,52 +107,165 @@ pub trait GpioPin: SealedPin + Sized + PeripheralType + Into<AnyPin> + 'static {
+            impl From<crate::peripherals::$peri> for AnyPin {
+                fn from(value: crate::peripherals::$peri) -> Self {
+                    value.degrade()
+                    }
```
slight formatting weirdness

MathisDeroo committed:
Create user-readable field for the pin configuration

Signed-off-by: Mathis Deroo <mathis.deroo@nxp.com>

jamesmunns commented:
@MathisDeroo could you please resolve the merge conflict on this?

jamesmunns commented on examples/src/bin/button.rs:
```diff
@@ -0,0 +1,35 @@
+
+    defmt::info!("Button example");
+
+   // Initialize embassy-time global driver backed by OSTIMER0
```
```suggestion
    // Initialize embassy-time global driver backed by OSTIMER0
```

jamesmunns commented on src/gpio.rs:
```diff
@@ -65,9 +140,23 @@ trait SealedPin {
 }
 
-/// GPIO pin trait.
+/// GP pin trait.
```
```suggestion
/// GPIO pin trait.
```

jamesmunns commented on src/gpio.rs:
```diff
@@ -395,6 +597,26 @@ impl<'d> Flex<'d> {
         !self.is_set_high()
     }
+
+    pub fn set_pull(&mut self, pull_select: Pull) {
```
For public functions, it is typical to include a `///` doc comment, which will be visible to users in the documentation view.

jamesmunns commented on src/gpio.rs:
```diff
@@ -404,10 +626,16 @@ pub struct Output<'d> {
+        initial: Level,
+        strength: DriveStrength,
+        slew_rate: SlewRate,
+        ) -> Self {
```
You might want to run `cargo fmt` on your PR to catch formatting issues like this! CI will check this, but it's not running due to the pending merge conflict.

jamesmunns reviewed (approved):
Overall looks good to me, a couple of minor nits, and the merge conflict needs to be fixed, but happy to get this merged ASAP!

Commit:
Merge remote-tracking branch 'odp/main' into gpio_support_v2

Commit:
examples: button: remove left-over OSTIMER initialization

While at that, also cargo fmt

felipebalbi commented:
> @MathisDeroo could you please resolve the merge conflict on this?

Just pushed a conflict fix and cargo fmt.

Commit:
Fix warnings

MathisDeroo committed:
Add documentation for public functions

Signed-off-by: Mathis Deroo <mathis.deroo@nxp.com>

felipebalbi commented on src/gpio.rs:
```diff
@@ -80,52 +168,157 @@ pub trait GpioPin: SealedPin + Sized + PeripheralType + Into<AnyPin> + 'static {
+    }
+
+    fn set_function(&self, function: Mux) {
+        self.pcr_reg().modify(|_, w| w.mux().variant(function));
```
modify will really read-modify-write, which is not atomic. Usually, we need to wrap modify calls within a `critical_section`. Ditto to all others.

```suggestion
        critical_section::with(|_| self.pcr_reg().modify(|_, w| w.mux().variant(function)));
```

MathisDeroo replied:
Should it be done to all functions, i.e private and public, and of all implementation or only impl SealedPin for AnyPin { as long as the modify is used?

felipebalbi replied:
actually, please ignore this comment. There's no situation where any of these would race.

felipebalbi commented on src/lib.rs:
```diff
@@ -188,7 +190,7 @@ embassy_hal_internal::peripherals!(
     P2_30,
     P2_31,
-
+    
```
unnecessary change

felipebalbi commented on src/gpio.rs:
```diff
@@ -80,52 +168,157 @@ pub trait GpioPin: SealedPin + Sized + PeripheralType + Into<AnyPin> + 'static {
+            3 => unsafe { &*crate::pac::Port3::ptr() },
+            4 => unsafe { &*crate::pac::Port4::ptr() },
+            _ => panic!("Invalid port: {}", self.port),
+        }
```
these already exist in the actual pin singleton, right? You can make the `AnyPin` counterpart simpler by holding the necessary register references within `AnyPin` struct, then just return them here and `pcr_reg()` below.

MathisDeroo replied:
Like adding to the AnyPin struct :
port_reg: &'static crate::pac::port0::RegisterBlock,
pcr_reg: &'static crate::pac::port0::Pcr0

And in the impl AnyPin { adding the two functions
fn port_reg(&self) -> &'static crate::pac::port0::RegisterBlock;{
        self.port
    }

fn pcr_reg(&self) -> &'static crate::pac::port0::Pcr0{
        self.pcr
    }

These two functions will be called in the impl SealedPin for AnyPin

felipebalbi replied:
Yup, what do you think?

MathisDeroo replied:
Looks way better ! Let me implement it

jamesmunns replied:
By the way @MathisDeroo, if you put your code in [code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks), it'll be a bit easier to read with formatting!

MathisDeroo committed:
Expose port and pcr registers to AnyPin implementation

Signed-off-by: Mathis Deroo <mathis.deroo@nxp.com>

MathisDeroo committed:
Remove unnecessary change

Signed-off-by: Mathis Deroo <mathis.deroo@nxp.com>

Commit:
Run cargo fmt

jamesmunns commented:
Thanks for all the iteration on getting this right @MathisDeroo! I appreciate it!

MathisDeroo commented:
Thanks @jamesmunns @felipebalbi for every comment and feedback ! 😄
